### ISSUE TITLE
After upgrading from 0.3.17 to 0.5.6, optimisticResponse does not work as expected.

### ISSUE THREAD

**Issue by HriBB** [apollostack/react-apollo#231]
I upgraded `react-apollo` from `0.3.17` to `0.5.6` and am rewriting components one by one for the new API. I noticed that some of my `optimisticResponse` mutations don't work as expected. It works if I change something simple, such as `active` from `true` to `false`, but not when a mutation is more complex and returns an array or deeply nested object.

For example, I have a `reorderBrands` mutation that accepts an array of IDs, sorted in the new order, and server returns a new list of `Brands`.

```
const REORDER_BRANDS = gql`
  mutation reorderBrands($ids: [Int!]!) {
    reorderBrands(ids: $ids) {
      id title description active locked pos
      category { id title }
    }
  }`

graphql(REORDER_BRANDS, {
  props: ({ ownProps, mutate }) => ({
    reorderBrands: (ids, brands) => mutate({
      variables: { ids },
      optimisticResponse: {
        reorderBrands: brands,
      },
    }),
  }),
}),
```

And this is how I call my mutation
```
reorder = (e, item, index, newIndex, list) => {
  const { reorderBrands } = this.props
  const ids = list.map(b => b.id)
  list.forEach((b, i) => b.pos = i + 1)
  reorderBrands(ids, list)
    .then(({ errors, data }) => {
      if (errors) console.log('==> errors', errors.message);
      if (data) console.log('==> data', data);
    })
}
```

I am probably doing something wrong, but I'm not sure what. It used to work with the `0.3.17` version. Any ideas? How can I debug this?

**Comment by HriBB** [comment]
I have also noticed that in this particular case, my list is not updated with the new result set from the server. Even if I remove `optimisticResponse`. Server returns a new array with the right order, but my list still shows the old order.

**Comment by HriBB** [comment]
I tried to debug this, but it's really hard. Not really sure if this is a problem with my code, `react-apollo` or `apollo-client`. It's crazy, because some simple `optimisticResponse`s work just fine, but not the code above, although the optimistic data and data returned from the server are exactly the same.

The only thing I have figured out so far: if I enable `optimisticResponse` in the code above, my component's `render` function is not called. Not for optimistic update and not after receiving data from the server. If I disable `optimisticResponse`, everything works just fine, but with a slight delay of course. 

@jbaxleyiii or @stubailo any ideas on how to debug this? Where should I even look?

**Comment by jbaxleyiii** [comment]
@HriBB I think the first step would be to try and either create a failing reproduction repo, or ideally, a failing test on react-apollo!

**Comment by HriBB** [comment]
So I finally figured this out. First of all, this is NOT a bug in `react-apollo`. There was a problem in my code, although I'm not really sure what I was doing wrong. I suspect I was mutating the existing data, and after I used `react-addons-update`, things started working again. I guess immutability comes with a price.

If anyone is interested in how I got it working, let me know ;) Closing this issue.
